nltk/lm/smoothing.py

```diff
@@ -72,66 +72,52 @@ def unigram_score(self, word):
 
     def unigram_score(self, word):
-        continuation_count, unique_continuation_count = self.continuation_counts(
-            (word,)
-        )
-        return continuation_count / unique_continuation_count
+        word_continuation_count, total_count = self._continuation_counts(word)
+        return word_continuation_count / total_count
 
     def alpha_gamma(self, word, context):
         prefix_counts = self.counts[context]
-        if len(context) + 1 == self._order:
-            prefix_total_ngrams = prefix_counts.N()
-            alpha = max(prefix_counts[word] - self.discount, 0.0) / prefix_total_ngrams
-            gamma = (
-                self.discount
-                * _count_non_zero_vals(prefix_counts)
-                / prefix_total_ngrams
-            )
-        else:
-            continuation_count, unique_continuation_count = self.continuation_counts(
-                context + (word,)
-            )
-            alpha = (
-                max(continuation_count - self.discount, 0.0) / unique_continuation_count
-            )
-            gamma = (
-                self.discount
-                * _count_non_zero_vals(prefix_counts)
-                / unique_continuation_count
-            )
+        word_continuation_count, total_count = (
+            (prefix_counts[word], prefix_counts.N())
+            if len(context) + 1 == self._order
+            else self._continuation_counts(word, context)
+        )
+        alpha = max(word_continuation_count - self.discount, 0.0) / total_count
+        gamma = self.discount * _count_non_zero_vals(prefix_counts) / total_count
         return alpha, gamma
 
-    def continuation_counts(self, context):
-        key_len = len(context) + 1
-        continuation_keys = []
-        unique_continuation_keys = []
-        if len(context) == 1:
-            key, value = None, context[-1]
-            for word in self.counts[2].keys():
-                unique_continuation_keys += [
-                    word + (sub_word,) for sub_word in self.counts[2][word]
-                ]
-                if context[-1] in self.counts[2][word].keys():
-                    continuation_keys.append(word + (value,))
-        else:
-            key, value = context[:-1], context[-1]
-            for ngram in self.counts[key_len].keys():
-                is_key = ngram[1:] == key
-                is_value = value in self.counts[key_len][ngram].keys()
-                if is_key:
-                    unique_continuation_keys += [
-                        ngram + (word,) for word in self.counts[key_len][ngram].keys()
-                    ]
-                    if is_value:
-                        continuation_keys.append(ngram + (value,))
-        continuation_count = len(set(continuation_keys))
-        unique_continuation_count = len(set(unique_continuation_keys))
-        return (continuation_count, unique_continuation_count)
+    def _continuation_counts(self, word, context=tuple()):
```

Comment:
A bit skeptical about this function. The idea is to find number of unique continuation. for eg. if the context is "ab" and c follows it 3 times, d follows 1, and e follows 1, the the unique continuation is only 3(c, d, and e), and not 5

Let me check this function a little more deeply and figure out if it works out the same or not.

Reply:
I'm taking advantage of the fact that all our frequency distributions already account for uniqueness when it comes to the keys. We never look at actual counts here.

Reply:
I guess you are right. The two functions are performing identically.

Reply:
Cool!

I think next weekend I'll try to come up with some test cases that might highlight the types vs tokens distinction.

Reply:
Past weekend ended up being hectic, I'll get back with the test examples (if any) on Thursday/Friday.

Reply:
After that this should be ready to merge I believe.

Reply:
Ok, I noticed that the comments explaining the tests were in need of some love, so I did that instead.

I think before I add more tests I should switch the whole setup to `pytest`, I'll do that in a separate PR. I already have a list of phenomena I want to test.

@manujosephv please double-check the accuracy of the new comments.

Reply:
Double checked the KneserNey, AbsoluteDiscounting, and StupidBackoff testcase comments, they seem to be correct.